docs/posts/1tbc/index.qmd

```diff
@@ -0,0 +1,733 @@
+        for f in filenames:
+            fp = os.path.join(dirpath, f)
+            total_size += os.path.getsize(fp)
+    return total_size
+
+
+sizes = [get_dir_size(table) for table in tables]
+names = [os.path.basename(table) for table in tables]
+
+tmp = ibis.memtable({"name": names, "size": sizes})
+tmp = tmp.mutate(size_gb=tmp["size"] / (1024**3))
+tmp = tmp.order_by(ibis.desc("size_gb"))
+tmp = tmp.mutate(size_gb_mem=tmp["size_gb"] * 11 / 5)
+
+c = px.bar(
+    tmp,
+    x="name",
+    y="size_gb",
+    title="table sizes in TPC-H data",
+    hover_data=["size_gb_mem"],
+    labels={
+        "name": "table name",
+        "size_gb": "size (GB on-disk in compressed Parquet files)",
+        "size_gb_mem": "size (approximate GB in memory)",
+    },
+    # color="size_gb",
+    # color_continuous_scale="viridis",
+)
+
+print(
+    f"total size: {tmp['size_gb'].sum().to_pyarrow().as_py():,.2f}GBs (compressed Parquet files)"
+)
+c
+```
+
+In-memory this would be about 1TB. Uncompressed CSV files would be >1TB on disk.
+
+Let's explore the largest table, `lineitem`. This table in memory is ~6X larger
+than RAM.
+
+```{python}
+#| code-fold: true
+#| code-summary: "Show code to explore the lineitem table"
+sf = 1
+n = 1
+data_dir = f"tpch_data/parquet/sf={sf}/n={n}"
+
+table_name = "lineitem"
+data = f"{data_dir}/{table_name}/*.parquet"
+
+t = ibis.read_parquet(data)
+print(f"rows: {t.count().to_pyarrow().as_py():,} | columns: {len(t.columns)}")
+```
+
+Over 6 billion rows!
+
+Let's try to display the first few rows with Ibis, pandas, and Polars:
+
+::: {.panel-tabset}
+
+## Ibis
+
+```{python}
+t = ibis.read_parquet(data)
+t.head(3)
+```
+
+## pandas
+
+```{.python}
+df = pd.concat([pd.read_parquet(f) for f in glob.glob(data)], ignore_index=True) # <1>
+df.head(3)
+```
+
+1. Work around lack of reading multiple parquet files in pandas
+
+```html
+The Kernel crashed while executing code in the current cell or a previous cell.
+Please review the code in the cell(s) to identify a possible cause of the failure.
+Click here for more info.
+View Jupyter log for further details.
+```
+
+## Polars (eager)
+
+```{.python}
+df = pl.read_parquet(data)
+df.head(3)
+```
+
+```html
+The Kernel crashed while executing code in the current cell or a previous cell.
+Please review the code in the cell(s) to identify a possible cause of the failure.
+Click here for more info.
+View Jupyter log for further details.
+```
+
+## Polars (lazy)
+
+```{python}
+df = pl.scan_parquet(data)
+df.head(3).collect()
+```
+
+## Polars (lazy, streaming)
+
+```{python}
+df = pl.scan_parquet(data)
+df.head(3).collect(streaming=True)
+```
+
+:::
+
+Ibis (DuckDB backend by default) can display the first few rows. Polars (lazy)
+can handle this in regular and streaming mode. pandas and Polars (eager) crash
+the kernel as they must load all the data into memory to construct their
+dataframes. Since this table is at ~6X larger than our 96GiB of RAM, we can't
```

Comment:
```suggestion
dataframes. Since this table is about ~6X larger than our 96GiB of RAM, we can't
```

Not sure if that should be `at least` perhaps?